theforeman.org/pipelines/release/foreman_packaging_deb_release.groovy

```diff
@@ -0,0 +1,94 @@
+def packages_to_build
+def build_steps
+
+pipeline {
+    agent { label 'debian' }
+
+    options {
+        timestamps()
+        timeout(time: 4, unit: 'HOURS')
+        disableConcurrentBuilds()
+        ansiColor('xterm')
+    }
+
+    stages {
+        stage('Clone Packaging') {
+            steps {
+                script {
+                    foreman_branch = foreman_version == 'nightly' ? "deb/develop" : "deb/${foreman_version}"
+                }
+
+                checkout([
+                    $class : 'GitSCM',
+                    branches : [[name: "*/${foreman_branch}"]],
+                    extensions: [[$class: 'CleanCheckout']],
+                    userRemoteConfigs: [
+                        [url: 'https://github.com/theforeman/foreman-packaging']
+                    ]
+                ])
+
+            }
+        }
+
+        stage('Find packages') {
+            steps {
+                copyArtifacts(projectName: env.JOB_NAME, optional: true)
+
+                script {
+
+                    if (fileExists('commit')) {
+
+                        commit = readFile(file: 'commit').trim()
+                        packages_to_build = find_changed_debs("${commit}..HEAD")
+                        build_steps = build_deb_package_steps(packages_to_build, foreman_version, 'theforeman', false)
+
+                    } else {
+
+                        build_steps
+
+                    }
+                }
+            }
+        }
+
+        stage('Setup sources') {
+            steps {
+                script {
+                    parallel build_steps.setup_sources
+                }
+            }
+        }
+
+        stage('Execute pbuilder') {
+            steps {
+                script {
+                    parallel build_steps.executes
+                }
+            }
+        }
+
+        stage('Stage packages') {
+            steps {
+                script {
+                    parallel build_steps.stage_packages
```

Comment:
I don't think this should be done in parallel. We rsync them over and then rebuild the archive. That's quite an expensive operation and if you can do it just once for multiple packages that's much cheaper.

Reply:
We already do this in parallel as I understand things. The original stage packages script is run per project-os combination. So a job like https://ci.theforeman.org/job/packaging_build_deb_dependency/559/console would run that script three times which could all happen at the same time? 

What rebuilds the archive? the rsync trigger that?

Reply:
On the server side there's a script that runs when rsync runs:
https://github.com/theforeman/foreman-infra/blob/master/puppet/modules/freight/templates/rsync.erb

As you can see, it runs the original command and then calls
```erb
  find <%= @home %>/rsync_cache/$DISTRO/$REPO -iname '*.deb' -exec freight-add -v -c <%= @home %>/freight.conf {} apt/$DISTRO/$REPO \;
  # Publish the debs
  freight-cache -c <%= @home %>/freight.conf -v apt/$DISTRO
  # Cleanup - no need to keep the debs
  find <%= @home %>/rsync_cache/$DISTRO/$REPO -iname '*.deb' -delete || true
```

Reply:
So what matters is `distro` and `repo`, so we could run this in parallel per OS and repository if we collected all those possible combinations up and constructed the run based on that?

Reply:
Sorry, what I meant to ask was, is there a model where we could initiate this script directly for distro/repo pairings based on what was built given that the rsync is based upon the actual projects that are built?

Reply:
Yes, we essentially need to gather all changed debs. Then make 2 lists:
* Those to staging (core + dependency)
* Those to release (plugin)

Then roughly `if (to_staging.any?) { rsync_to_staging(to_staging) }` and the same with `to_release`.

You're right that this is an optimization we didn't do before, but can do now.

Reply:
Is there anything that would prevent us from moving to a model where all of or parts of https://github.com/theforeman/foreman-infra/blob/master/puppet/modules/freight/templates/rsync.erb are initiated explicitly rather than implicitly due to an rsync? I am thinking that would allow us to have cleaner optimized code if the archive building is the expensive operation.